### DISCUSSION TITLE
Express Backend Deployment gives ERROR 404_NOT_FOUND

### DISCUSSION THREAD

**Discussion by SengulC** [vercel/community#6530] [Help]
### Summary

My express backend works locally, but once i deploy on vercel i get the 404_NOT_FOUND error. I don't know what to do. 

My backend file (index.js) and package/vercel .jsons are all under /api folder. I have the following imports in index.js, could those be causing the issue? 

```
var express = require("express");
var cors = require("cors");
var OpenAI = require("openai");
var bodyParser = require("body-parser");
```

But my deployment page says the index.js serverless function was found.

### Steps to Reproduce

```
# package.json:
{
  "name": "server",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "nodemon index.js",
    "build": "esbuild index.js --bundle --platform=node --target=node14 --outfile=dist/index.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "axios": "^1.6.7",
    "cors": "^2.8.5",
    "express": "^4.18.2"
  },
  "devDependencies": {
    "esbuild": "^0.20.2",
    "nodemon": "^3.0.3"
  }
}


# vercel.json
{
  "version": 2,
  "builds": [
    {
      "src": "index.js",
      "use": "@vercel/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "index.js"
    }
  ]
}

# logs
Running build in Washington, D.C., USA (East) – iad1
Cloning github.com/SengulC/guai-0.5 (Branch: main, Commit: b340c29)
Cloning completed: 1.085s
Restored build cache
Running "vercel build"
Vercel CLI 33.7.1
Installing dependencies...
added 108 packages in 4s
16 packages are looking for funding
  run `npm fund` for details
Build Completed in /vercel/output [5s]
Deploying outputs...
Deployment completed
Uploading build cache [5.02 MB]...
Build cache uploaded: 894.779ms

```

**Comment by amyegan** [comment]
Hey @SengulC. I don't know enough about your project to say for sure, but I suspect the file structure or configuration may be the cause of problem. I recommend following our guide on using Express.js with Vercel: https://vercel.com/guides/using-express-with-vercel

I used an older version of that guide to build my express sample app. In case it helps to compare with another example, the repo is public: https://github.com/amyegan/express-vercel

**Reply by SengulC** [reply]
Hi @amyegan, I've actually reviewed both the guide and your public repo to help me but have changed nothing ':(. Here's my file structure, the config files are in my original post.
```
- root
  - server
     - api
        - index.js
        - package.json
        - package-lock.json
        - vercel.json   
```
And here's the essential part of my Express app:
```
const app = express();
app.use(bodyParser.json());
app.use(cors());

app.post("/buttons", async (req, res) => {
  const input = req.body.nodelabel;
  const prompt = req.body.prompt;
  const brief = req.body.brief;
  const nodes = req.body.nodes;
  let result = await callButtonPrompt(prompt, input, brief, nodes);
  result = result.message.content; 
  res.send(result);
});
const port = process.env.PORT || 8000;

// Landing Page
app.get('/', (req, res) => {
  res.send('Server is running');
});

app.listen(port, () => {
  console.log(`Server is running on port ${port}.`);
});
```

Note: i use npm start to run my server locally.

**Reply by SengulC** [reply]
One last reply/edit sorry, I tried moving the files to the same structure you had it (so only index.js in the /api folder), i still get 404 NOT_FOUND but now i don't even get a deployment summary marking the index.js function. I'm so confused.

**Reply by amyegan** [reply]
Do you have `server` set as the [Root Directory](https://vercel.com/docs/deployments/configure-a-build#root-directory) in project settings?

**Reply by SengulC** [reply]
Yes
